### PR TITLE
[LSP] Fix fading diagnostics

### PR DESCRIPTION
**Pull request by dibarbet**
The logic for fading diagnostics changed to use additional locations instead of always creating multiple diagnostics with different spans.  This updates LSP to read the new diagnostic data and itself create multiple diagnostic reports for the unnecessary locations.  LSP does not yet support having additional locations with different tags.

LSP diagnostics should now appropriately respect severities and tools options for fading.  Also added tests to verify LSP behavior.

In the screenshot, usings and unreachable code are fully faded.  Unnecessary parens fades the parens and warns on the first line when configured to warning severity.
image

Resolves https://github.com/dotnet/roslyn/issues/63147

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticDataExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+using System.Collections.Immutable;
+using System.IO;
+using System.Runtime.Serialization.Json;
+using System.Text;
+using System;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Features.Diagnostics;
+internal static class DiagnosticDataExtensions
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticDataExtensions.cs:16]: this code was just moved into a common layer out of the tagger.

File: src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs [hunk 2]
```diff
@@ -391,13 +413,21 @@ protected static DiagnosticTag[] ConvertTags(DiagnosticData diagnosticData, bool
...
                 result.Add(VSDiagnosticTags.EditAndContinueError);
 
             return result.ToArray();
         }
+
+        private bool SupportsFadingOption(DiagnosticData diagnosticData)
+        {
+            if (IDEDiagnosticIdToOptionMappingHelper.TryGetMappedFadingOption(diagnosticData.Id, out var fadingOption))
+            {
+                return diagnosticData.Language != null
```
Comment (Youssef1313) [inline src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:426]: Do we have a case where this happens?

**Comment by mavasani** [comment]
@dibarbet So the approach in general looks good to me, and matches what we discussed offline couple of weeks ago. However, I am wondering if it is simpler to just remove this encoding logic in all such fading analyzers and instead have the analyzers themselves report separate hidden, non-configurable, unnecessary tagged diagnostics, maybe with a separate diagnostic ID. This way we avoid this unnecessary round-trip. I don't recall now why we originally decided to move from separate fading diagnostics to this encoding logic in additional locations - maybe design wise it is better to have a single diagnostic with additional locations for fading, but it seems performance and maintenance wise it might be better to go with separate diagnostics in the analyzer. What do you think?

File: src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs [hunk 3]
```diff
@@ -273,37 +271,54 @@ private void HandleRemovedDocuments(RequestContext context, ImmutableArray<Previ
...
-                };
+                return ImmutableArray.Create<LSP.Diagnostic>(diagnostic);
+            }
 
-                return vsDiagnostic;
+            // Check to see if there are specific locations marked to fade.
+            if (!diagnosticData.TryGetUnnecessaryDataLocations(out var unnecessaryLocations))
+            {
+                // There are no specific fading locations, just mark the whole diagnostic span as unnecessary.
+                diagnostic.Tags = diagnostic.Tags?.Append(DiagnosticTag.Unnecessary);
```
Comment (mavasani) [inline src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:295]: Can `Tags` really be null here?

File: src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs [hunk 4]
```diff
@@ -273,37 +271,54 @@ private void HandleRemovedDocuments(RequestContext context, ImmutableArray<Previ
...
-            // We can just use VSDiagnostic as it doesn't have any default properties set that
-            // would get automatically serialized.
-            LSP.VSDiagnostic CreateBaseLspDiagnostic()
+            // Roslyn produces unnecessary diagnostics by using additional locations, however LSP doesn't support tagging
+            // additional locations separately.  Instead we just create multiple hidden diagnostics for unnecessary squiggling.
+            using var _ = ArrayBuilder<LSP.Diagnostic>.GetInstance(out var diagnosticsBuilder);
+            diagnosticsBuilder.Add(diagnostic);
+            foreach (var location in unnecessaryLocations)
             {
+                var additionalDiagnostic = CreateLspDiagnostic(diagnosticData, project, capabilities);
```
Comment (mavasani) [inline src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:305]: Do we also need to change the Severity to Hidden for this additionalDiagnostic?

**Review by mavasani — APPROVED** [review]

**Comment by Youssef1313** [comment]
> I don't recall now why we originally decided to move from separate fading diagnostics to this encoding logic in additional locations

@mavasani A bit of history here. This was introduced in https://github.com/dotnet/roslyn/pull/39649. There is no associated issues to the PR, but branch name is "dupe_error_list". We were using same ID for both descriptors `s_diagnosticWithFade` and `s_diagnosticWithoutFade` and both were **configurable**, so I think this can indeed lead to duplicate diagnostics. Using `isConfigurable: false` to one of these descriptors and changing its ID should be good to fix the duplicate diagnostics in a more performant way as you suggested.

One possible downside is that now these descriptors will appear in EditorConfig UI, but it's not much important I think.

**Comment by dibarbet** [comment]
> maybe design wise it is better to have a single diagnostic with additional locations for fading, but it seems performance and maintenance wise it might be better to go with separate diagnostics in the analyzer.

IMHO the right solution is to have LSP support tags on additional locations.  The problem with multiple diagnostics is its pretty easy to accidentally start showing them in places they aren't supposed to be, e.g. duplicate hover info, duplicate error list info, configuring hidden diagnostics accidentally, etc.  

LSP already has the concept of diagnostic related information, I think we just need to extend it to add tags - https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticRelatedInformation

File: src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs (hunk 3 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:295]: no, let me assert that

**Comment by dibarbet** [comment]
Forgot to mention one thing as well - VSCode doesn't have as complex support for diagnostics as VS does and doesn't have diagnostics that can be 'hidden'.  IMHO it will be much easier to expand tags to additional diagnostic locations that it will be to introduce the concept of a hidden diagnostic in VSCode.

File: src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs (hunk 2 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:426]: hmm I don't think so. will assert

File: src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs (hunk 4 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs:305]: update - LSP doesn't have a hidden severity, instead the visibility is controlled by the tags.  I changed it to the lowest severity anyway

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticDataExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+using Microsoft.CodeAnalysis.ErrorReporting;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Runtime.Serialization.Json;
+using System.Text;
+using System;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticDataExtensions.cs:13]: Sort :)

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticDataExtensions.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/DiagnosticDataExtensions.cs:16]: nit: i'd prfer a space above here.

**Pull request merged**